lib/oli_web/live/experiments/experiments_live.ex

```diff
@@ -26,8 +47,423 @@ defmodule OliWeb.Experiments.ExperimentsView do
         name="experiments"
         value={nil}
         label="Enable A/B testing with UpGrade"
+        phx-click="enable_upgrade"
+        checked={@is_upgrade_enabled}
       />
+
+      <%= if @experiment do %>
+        <OliWeb.Resources.AlternativesEditor.group
+          group={@experiment}
+          editing_enabled={@is_upgrade_enabled}
+          source={:experiments}
+        />
+      <% end %>
     </div>
     """
   end
+
+  def handle_event("enable_upgrade", params, socket) do
+    socket =
+      case socket.assigns.experiment do
+        nil ->
+          {:ok, experiment} =
+            create_experiment(socket.assigns.project, socket.assigns.current_author)
+
+          assign(socket, experiment: experiment, is_upgrade_enabled: true)
+
+        %Revision{} ->
+          if Map.get(params, "value") do
+            assign(socket, :is_upgrade_enabled, true)
+          else
+            assign(socket, :is_upgrade_enabled, false)
+          end
+      end
+
+    {:noreply, socket}
+  end
+
+  def handle_event("show_create_option_modal", %{"resource_id" => resource_id}, socket) do
+    changeset =
+      {%{id: uuid(), resource_id: resource_id}, %{id: :string, resource_id: :int, name: :string}}
+      |> Ecto.Changeset.cast(%{}, [:id, :resource_id, :name])
+
+    form_body_fn = fn assigns ->
+      ~H"""
+      <div class="form-group">
+        <%= hidden_input(@form, :id) %>
+        <%= hidden_input(@form, :resource_id) %>
+
+        <%= text_input(
+          @form,
+          :name,
+          class: "form-control my-2" <> error_class(@form, :name, "is-invalid"),
+          placeholder: "Enter a name",
+          phx_hook: "InputAutoSelect",
+          required: true
+        ) %>
+      </div>
+      """
+    end
+
+    modal_assigns = %{
+      id: "create_modal",
+      title: "Create Option",
+      submit_label: "Create",
+      changeset: changeset,
+      form_body_fn: form_body_fn,
+      on_validate: "validate_option",
+      on_submit: "create_option"
+    }
+
+    modal = fn assigns ->
+      ~H"""
+      <FormModal.modal {@modal_assigns} />
+      """
+    end
+
+    {:noreply, show_modal(socket, modal, modal_assigns: modal_assigns)}
+  end
+
+  def handle_event(
+        "create_option",
+        %{"params" => %{"id" => option_id, "name" => name, "resource_id" => resource_id}},
+        socket
+      ) do
+    ###
+    resource_id = ensure_integer(resource_id)
+
+    %{content: %{"options" => options} = content} = socket.assigns.experiment
+
+    new_options = [%{"id" => option_id, "name" => name} | options]
+
+    %{project: project, ctx: ctx} = socket.assigns
```

Comment:
what do you think about getting the `options`, the `project` and the `ctx` in one single pattern matching with `socket.assigns`?

Reply:
I changed for this format, otherwise the extraction is going to look odd.


```
%{project: project, ctx: ctx, experiment: experiment} = socket.assigns
%{content: %{"options" => options} = content} = experiment
new_options = [%{"id" => option_id, "name" => name} | options]
```